### PR TITLE
Add CAN

### PR DESCRIPTION
**Pull request by auscompgeek**
~~On top of #44 and #46. Sorry not sorry.~~

Opening for visibility and feedback. `wpilib/src/can.rs` is the important part.

I have a WIP [port of the SPARK MAX libraries](https://gitlab.com/auscompgeek/sparkmax-rs) that uses this.

### CODE REVIEW

**Comment by Lytigas** [comment]
Rebase at some point please.

File: wpilib-sys/src/hal_call.rs [hunk 1]
```diff
@@ -124,13 +118,14 @@ impl<T> HalMaybe<T> {
 /// the `Err` variant.
 #[macro_export]
 macro_rules! hal_call {
-    ($function:ident($($arg:expr),*)) => {{
+    ($function:ident($($arg:expr,)*)) => {{
```
Comment (Lytigas) [inline wpilib-sys/src/hal_call.rs]: I'm not quite sure how this is different from the other pattern.
https://doc.rust-lang.org/reference/macros-by-example.html shows patterns declared like the old version.
Right now my guess is that this case requires a trailing comma, and the other requires no trailing comma. If that's the case, adding a comment might be helpful.

File: wpilib/src/can.rs [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+
+impl Can {
+    pub const TEAM_MANUFACTURER: HAL_CANManufacturer::Type = TEAM_MANUFACTURER;
+    pub const TEAM_DEVICE_TYPE: HAL_CANDeviceType::Type = TEAM_DEVICE_TYPE;
+
+    pub fn new(
+        device_id: i32,
+        device_manufacturer: HAL_CANManufacturer::Type,
+        device_type: HAL_CANDeviceType::Type,
+    ) -> HalResult<Self> {
```
Comment (Lytigas) [inline wpilib/src/can.rs]: Per API guidelines (https://rust-lang-nursery.github.io/api-guidelines/naming.html), I think `new` is supposed to have fewer arguments. However, if `TEAM_MANUFACTURER` refers to a CAN device made _by a team_, then I guess that is the far less common case, and there can't really be a good default for `new`. Not sure what to do about that.

File: wpilib/src/can.rs [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+        let mut timestamp = 0;
+        let status = hal_call!(HAL_ReadCANPacketNew(
+            self.handle,
+            api_id,
+            data.as_mut_ptr(),
+            &mut length,
+            &mut timestamp,
+        ));
+        match status {
+            Err(HalError(code)) if code == HAL_ERR_CANSessionMux_MessageNotFound => Ok(None),
```
Comment (Lytigas) [inline wpilib/src/can.rs]: To clarify: This handles the case where the read succeeds, but there's just no data, right?

Comment (Lytigas) [reply wpilib/src/can.rs]: Would this not be more readable as `Err(HalError(HAL_ERR_CANSessionMux_MessageNotFound))`?

Comment (auscompgeek) [reply wpilib/src/can.rs]: > To clarify: This handles the case where the read succeeds, but there's just no data, right?

I believe this is the "everything you gave me is valid, but I haven't seen any frames for you" condition.

> Would this not be more readable as `Err(HalError(HAL_ERR_CANSessionMux_MessageNotFound))`?

I believe that actually means "unwrap the HalError and put its value in a variable `HAL_ERR_CANSessionMux_MessageNotFound`". Feel free to correct me if I'm wrong.

Comment (Lytigas) [reply wpilib/src/can.rs]: > put its value in a variable HAL_ERR_CANSessionMux_MessageNotFound

The match compares the constant value. See [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=2516975896543bac45112051ace2c48c).

Comment (auscompgeek) [reply wpilib/src/can.rs]: Hm. I did get a rustc lint warning from doing it that way... Ah, current Rust stable gives a more descriptive warning. See [this playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=50107d4896d917d538f640c78a3f8ce0).

I'm tempted to keep it like this to mirror the other methods anyway.

File: wpilib-sys/src/hal_call.rs (hunk 1 above)
Comment (auscompgeek) [reply wpilib-sys/src/hal_call.rs]: If this were using Rust 2018, we could consolidate the two patterns into one (without allowing an arbitrary number of trailing commas). Alas, this is not 2018 edition.

Comment (Lytigas) [reply wpilib-sys/src/hal_call.rs]: I can't believe I forgot to update to 2018 edition! I'll start work on that, and then with a rebase we could remove the extra pattern. Unless there's a reason to stay on 2015?

File: wpilib/src/can.rs (hunk 3 above)
Comment (Lytigas) [reply wpilib/src/can.rs]: I was thinking change the other methods to inline style as well, I just only marked this one to save space. However, the compiler does bring up a good point. A non-upper-case global in a match could be confusing to readers, so I guess it makes sense to keep it as is.

File: wpilib-sys/src/hal_call.rs (hunk 1 above)
Comment (auscompgeek) [reply wpilib-sys/src/hal_call.rs]: Sounds good.

Comment (Lytigas) [reply wpilib-sys/src/hal_call.rs]: Done!

**Comment by Lytigas** [comment]
Other than the rebase and macro change for 2018 edition, what else needs to be done? Anything I could work on?

**Comment by auscompgeek** [comment]
Probably also worth adding docs on the CanData methods.

**Comment by auscompgeek** [comment]
Finally got around to updating this - have fun.

**Pull request merged**